File: leetcode/1696. Jump Game VI.py
class Solution:
    def maxResult(self, nums: List[int], k: int) -> int:
        res = [-sys.maxsize] * len(nums)
        res[0] = nums[0]
        dq = deque()
        for i in range(len(nums)):
            while len(dq) != 0 and i > dq[0]+k:
                dq.popleft()
            if len(dq) != 0:
                res[i] = max(res[i], res[dq[0]]+nums[i])
            while len(dq) != 0 and res[i] >= res[dq[-1]]:
                dq.pop()
            dq.append(i)
        return res[-1]

File: leetcode/1696. Jump Game VI.rs
use std::cmp;

impl Solution {
    pub fn max_result(nums: Vec<i32>, k: i32) -> i32 {
        let k = k as usize;
        let mut dp = vec![0; nums.len()];
        dp[0] = nums[0];
        
        for i in (1..nums.len()) {
            dp[i] = i32::MIN;
            let mut j = i - 1;
            while j + k >= cmp::max(i, k) {
                dp[i] = cmp::max(dp[i], dp[j]);
                j -= 1;
            }
            dp[i] += nums[i]
        }
        
        dp[nums.len()-1]
    }
}
      
      
